File: java/src/main/java/money_problem/domain/Portfolio.java
package money_problem.domain;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public final class Portfolio {
    private final List<Money> moneys;

    public Portfolio() {
        this.moneys = new ArrayList<>();
    }

    private Portfolio(List<Money> moneys) {
        this.moneys = Collections.unmodifiableList(moneys);
    }

    public Portfolio add(Money money) {
        var updatedMoneys = new ArrayList<>(moneys);
        updatedMoneys.add(money);

        return new Portfolio(updatedMoneys);
    }

    public Money evaluateWithException(Bank bank, Currency toCurrency) throws MissingExchangeRatesException {
        var convertedMoneys = convertAllMoneys(bank, toCurrency);

        if (containsFailure(convertedMoneys)) {
            throw toMissingExchangeRatesException(convertedMoneys);
        }
        return toMoney(convertedMoneys, toCurrency);
    }

    private boolean containsFailure(List<ConversionResult> convertedMoneys) {
        return convertedMoneys
                .stream()
                .anyMatch(ConversionResult::isFailure);
    }

    private List<ConversionResult> convertAllMoneys(Bank bank, Currency toCurrency) {
        return moneys.stream()
                .map(money -> convertMoney(bank, money, toCurrency))
                .toList();
    }

    private MissingExchangeRatesException toMissingExchangeRatesException(List<ConversionResult> convertedMoneys) {
        return new MissingExchangeRatesException(
                convertedMoneys.stream()
                        .filter(ConversionResult::isFailure)
                        .map(ConversionResult::missingExchangeRateException)
                        .toList()
        );
    }

    private Money toMoney(List<ConversionResult> convertedMoneys, Currency toCurrency) {
        return new Money(convertedMoneys.stream()
                .filter(ConversionResult::isSuccess)
                .mapToDouble(c -> c.money.amount())
                .sum(), toCurrency);
    }

    private ConversionResult convertMoney(Bank bank, Money money, Currency toCurrency) {
        try {
            return new ConversionResult(bank.convert(money, toCurrency));
        } catch (MissingExchangeRateException missingExchangeRateException) {
            return new ConversionResult(new MissingExchangeRatesException(missingExchangeRateException));
        }
    }

    public ConversionResult evaluate(Bank bank, Currency usd) {
        try {
            return new ConversionResult(evaluateWithException(bank, usd));
        } catch (MissingExchangeRatesException e) {
            return new ConversionResult(e);
        }

    }

    private record ConversionResult(Money money, MissingExchangeRatesException missingExchangeRateException) {
        public ConversionResult(Money money) {
            this(money, null);
        }

        public ConversionResult(MissingExchangeRatesException missingExchangeRateException) {
            this(null, missingExchangeRateException);
        }

        public boolean isFailure() {
            return missingExchangeRateException != null;
        }

        public boolean isSuccess() {
            return money != null;
        }

        public Money success() {
            return money();
        }
    }
}

File: c#/money-problem.Domain/Portfolio.cs
using System.Collections.Immutable;

namespace money_problem.Domain;

public class Portfolio
{
    private readonly ICollection<Money> moneys;

    public Portfolio()
    {
        this.moneys = new List<Money>();
    }

    private Portfolio(IEnumerable<Money> moneys)
    {
        this.moneys = moneys.ToImmutableList();
    }

    private static MissingExchangeRatesException ToException(IEnumerable<ConversionResult<MissingExchangeRateException>> results) =>
        new(results
            .Where(result => result.IsFailure())
            .Select(result => result.GetFailureUnsafe())
            .ToList());

    private static Money ToMoney(IEnumerable<ConversionResult<MissingExchangeRateException>> results, Currency currency) =>
        new(results.Sum(result => result.GetSuccessUnsafe().Amount), currency);

    private static bool ContainsFailure(IEnumerable<ConversionResult<MissingExchangeRateException>> results) =>
        results.Any(result => result.IsFailure());

    private List<ConversionResult<MissingExchangeRateException>> GetConvertedMoneys(Bank bank, Currency currency) =>
        this.moneys
            .Select(money => ConvertMoney(bank, currency, money))
            .ToList();

    private static ConversionResult<MissingExchangeRateException> ConvertMoney(Bank bank, Currency currency, Money money)
    {
        try
        {
            return new ConversionResult<MissingExchangeRateException>(bank.Convert(money, currency));
        }
        catch (MissingExchangeRateException exception)
        {
            return new ConversionResult<MissingExchangeRateException>(exception);
        }
    }

    public Portfolio Add(Money money)
    {
        List<Money> updatedMoneys = this.moneys.ToList();
        updatedMoneys.Add(money);
        return new Portfolio(updatedMoneys);
    }

    public ConversionResult<string> Evaluate(Bank bank, Currency currency)
    {
        List<ConversionResult<MissingExchangeRateException>> results = this.GetConvertedMoneys(bank, currency);
        return ContainsFailure(results)
            ? new ConversionResult<string>(ToException(results).Message)
            : new ConversionResult<string>(ToMoney(results, currency));
    }
}